Port WebExtension Display Strings to C++

ItsJamie9494 opened the pull request:
#### fdd79f3b8daaecb342721c8103d55f129b80ee38
<pre>
Port WebExtension Display Strings to C++
<a href="https://webkit.org/b/279610">https://webkit.org/b/279610</a>

Reviewed by Timothy Hatcher.

Create a JSON manifest to use in parallel with the Cocoa NSDictionary, and port the Display Strings to C++ (version, name, description).

* Source/WTF/wtf/JSONValues.h:
(WTF::JSONImpl::Value::operator! const):
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension):
(WebKit::WebExtension::parseManifest):
(WebKit::WebExtension::manifestParsedSuccessfully): Deleted
(WebKit::WebExtension::manifestVersion): Deleted.
(WebKit::WebExtension::displayName): Deleted.
(WebKit::WebExtension::displayShortName): Deleted.
(WebKit::WebExtension::displayVersion): Deleted.
(WebKit::WebExtension::displayDescription): Deleted.
(WebKit::WebExtension::version): Deleted.
(WebKit::WebExtension::populateDisplayStringsIfNeeded): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp: Added.
(WebKit::WebExtension::manifestParsedSuccessfuly):
(WebKit::WebExtension::manifestObject):
(WebKit::WebExtension::manifestVersion):
(WebKit::WebExtension::displayName):
(WebKit::WebExtension::displayShortName):
(WebKit::WebExtension::displayVersion):
(WebKit::WebExtension::displayDescription):
(WebKit::WebExtension::version):
(WebKit::WebExtension::populateDisplayStringsIfNeeded):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:

Canonical link: <a href="https://commits.webkit.org/283662@main">https://commits.webkit.org/283662@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c499e2f6c478f5d7470cc0316eae77e0c4281c0e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66947 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46322 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19569 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70982 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18080 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54121 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17861 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/53707 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12162 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70014 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42608 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57922 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/34228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39279 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15316 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16434 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/60061 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61198 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15657 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72683 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/66192 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10904 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15000 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/61181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10936 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57981 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61256 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8955 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2569 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/87960 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10164 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/42129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15480 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43206 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44389 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42949 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -237,6 +233,11 @@
 
+    auto manifestJSONString = String([[NSString alloc] initWithData:manifestData encoding:NSUTF8StringEncoding]);
+    m_manifestJSON = JSON::Value::parseJSON(manifestJSONString);
+    if (!m_manifestJSON)
```
`m_manifestJSON` is a `Ref`, so the result of `JSON::Value::parseJSON()` must be checked for `nullptr` first. If parsing succeeds, assign `*result` to `m_manifestJSON`; otherwise, assign `JSON::Value::null()`. This mismatch is the cause of the build failure.

xeenon replied:
You might want to move over `manifestParsedSuccessfully()` in this PR, and check `m_manifestJSON` for `isNull()`. Just for completeness on the parsing part.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3270,7 +3270,7 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 {
 ALLOW_NONLITERAL_FORMAT_BEGIN
-    return [NSString localizedStringWithFormat:WEB_UI_STRING("%@ Web Extension", "Extension's process name that appears in Activity Monitor where the parameter is the name of the extension"), extension().displayShortName()];
+    return [NSString localizedStringWithFormat:WEB_UI_STRING("%@ Web Extension", "Extension's process name that appears in Activity Monitor where the parameter is the name of the extension"), extension().displayShortName().span8()];
```
The `%@` format specifier requires a NS/CF object. Just cast to `NSString`:

```suggestion
    return [NSString localizedStringWithFormat:WEB_UI_STRING("%@ Web Extension", "Extension's process name that appears in Activity Monitor where the parameter is the name of the extension"), static_cast<NSString *>(extension().displayShortName())];
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -0,0 +1,134 @@
+    auto value = manifestObject->getDouble(manifestVersionManifestKey);
+
+    if (value)
+        return *value;
```
```suggestion
    if (auto value = manifestObject->getDouble(manifestVersionManifestKey))
        return *value;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -216,7 +211,8 @@
 
 WebExtension::WebExtension(NSDictionary *resources)
-    : m_resources([resources mutableCopy] ?: [NSMutableDictionary dictionary])
+    : m_manifestJSON(JSON::Object::create())
```
`JSON::Value::null()` is likely better before we parse the manifest.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -205,7 +199,8 @@
 
 WebExtension::WebExtension(NSDictionary *manifest, NSDictionary *resources)
-    : m_resources([resources mutableCopy] ?: [NSMutableDictionary dictionary])
+    : m_manifestJSON(JSON::Object::create())
```
Ditto on `null()`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -190,6 +183,7 @@
 
 WebExtension::WebExtension(NSURL *resourceBaseURL, NSError **outError)
     : m_resourceBaseURL(resourceBaseURL.URLByStandardizingPath.absoluteURL)
+    , m_manifestJSON(JSON::Object::create())
```
Ditto on `null()`.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -0,0 +1,134 @@
+{
+    auto manifestObject = m_manifestJSON->asObject();
+    if (!manifestObject)
+        return 0; // 0 is not a valid manifest version
```
```suggestion
        return 0;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -0,0 +1,134 @@
+
+    m_displayName = manifestObject->getString(nameManifestKey);
+
+    m_displayShortName = manifestObject->getString(shortNameManifestKey);
```
```suggestion
    m_displayName = manifestObject->getString(nameManifestKey);
    m_displayShortName = manifestObject->getString(shortNameManifestKey);
```

xeenon reviewed (changes requested):
This is much easier to review. Thanks again for doing this!

Looks good once you get the build failure resolved.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -237,6 +225,13 @@
+    if (!manifestJSON)
+        recordError(createError(Error::InvalidManifest, nil, nil));
+    else
+        m_manifestJSON = *manifestJSON;
```
This needs to return `false` on error, and should do the `asObject()` check.

```suggestion
    if (!manifestJSON || !manifestJSON->asObject()) {
        recordError(createError(Error::InvalidManifest, nil, nil));
        return false;
    }

    m_manifestJSON = *manifestJSON;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -0,0 +1,148 @@
+        m_manifestJSON = JSON::Value::null();
+    }
+
+    return isManifestValid;
```
The `asObject()` check should be in the `parseManifest()` method. I moved it in my suggestion. The `createError()` call was also missing a call to `recordError()`.

```suggestion
    // If we haven't parsed yet, trigger a parse by calling the getter.
    return !!manifestObject();
```

Also adopts `manifestObject()` from my other suggestion, to move away from the `NSDictionary` version of `manifest()`.

ItsJamie9494 replied:
by doing this, the NSDictionary copy of manifest is not parsed, since manifest() is never called (except in some error code it looks like). Could something like `!!manifest() && !!manifestObject()` work?

ItsJamie9494 replied:
also by not calling manifest(), m_parsedManifest is never set, which means manifestParsedSuccessfully will call manifestObject, which will continue to call manifestParsedSuccessfully in a loop.

xeenon replied:
Good point. Yeah checking both is fine!

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -0,0 +1,148 @@
+{
+    auto manifestObject = m_manifestJSON->asObject();
+    if (!manifestObject)
+        return 0;
```
This will currently fail unless something else parses the manifest first. This should use `manifestParsedSuccessfully()`.

```suggestion
    if (!manifestParsedSuccessfully())
        return 0;

    RefPtr manifestObject = m_manifestJSON->asObject();
    ASSERT(manifestObject);
```

I also don't like the boilerplate it adds. Maybe we want a small method that wraps this up a little tighter? Adding a new `manifestObject()` getter would be helpful.

```c++
RefPtr<JSON::Object> manifestObject() const { return manifestParsedSuccessfully() ? m_manifestJSON->asObject() : nullptr; }
```

Then:
```suggestion
    RefPtr manifestObject = this->manifestObject();
    if (!manifestObject)
        return 0;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -0,0 +1,148 @@
+bool WebExtension::manifestParsedSuccessfully()
+{
+    if (m_parsedManifest)
+        return !!m_manifest;
```
```suggestion
        return !!m_manifestJSON->asObject();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -0,0 +1,148 @@
+    m_parsedManifestDisplayStrings = true;
+
+    auto manifestObject = m_manifestJSON->asObject();
+    ASSERT(manifestObject);
```
```suggestion
    if (m_parsedManifestDisplayStrings)
        return;

    m_parsedManifestDisplayStrings = true;

    RefPtr manifestObject = this->manifestObject();
    if (!manifestObject)
        return;
```

ItsJamie9494 committed:
Port WebExtension Display Strings to C++
https://webkit.org/b/279610

Reviewed by Timothy Hatcher.

Create a JSON manifest to use in parallel with the Cocoa NSDictionary, and port the Display Strings to C++ (version, name, description).

* Source/WTF/wtf/JSONValues.h:
(WTF::JSONImpl::Value::operator! const):
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension):
(WebKit::WebExtension::parseManifest):
(WebKit::WebExtension::manifestParsedSuccessfully): Deleted
(WebKit::WebExtension::manifestVersion): Deleted.
(WebKit::WebExtension::displayName): Deleted.
(WebKit::WebExtension::displayShortName): Deleted.
(WebKit::WebExtension::displayVersion): Deleted.
(WebKit::WebExtension::displayDescription): Deleted.
(WebKit::WebExtension::version): Deleted.
(WebKit::WebExtension::populateDisplayStringsIfNeeded): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp: Added.
(WebKit::WebExtension::manifestParsedSuccessfuly):
(WebKit::WebExtension::manifestObject):
(WebKit::WebExtension::manifestVersion):
(WebKit::WebExtension::displayName):
(WebKit::WebExtension::displayShortName):
(WebKit::WebExtension::displayVersion):
(WebKit::WebExtension::displayDescription):
(WebKit::WebExtension::version):
(WebKit::WebExtension::populateDisplayStringsIfNeeded):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:

Canonical link: https://commits.webkit.org/283662@main
